loopy/kernel/function_interface.py

```diff
@@ -303,21 +322,23 @@ class AbstractExpressionToCodeMapper(Protocol):
     .. automethod:: __call__
     """
 
-    def infer_type(self, expr: Expression) -> LoopyType:
-        ...
+    def infer_type(self, expr: Expression) -> LoopyType: ...
 
     def __call__(self,
                  expr: Expression,
                  prec: int | None = None,
                  type_context: str | None = None,
-                 needed_dtype: LoopyType | None = None
-             ) -> GeneratedExpression:
-        ...
+                 needed_dtype: LoopyType | None = None) -> GeneratedExpression: ...
+
+    def rec(self,
+            expr: Expression,
+            type_context: str | None = None,
+            needed_dtype: LoopyType | None = None) -> GeneratedExpression: ...
```

Comment:
This `rec` is used in `ScalarCallable.emit_call`, but it doesn't seem to have 4 arguments like `__call__`. Is that correct?